runtime/in-memory-storage.js

```diff
@@ -17,12 +17,17 @@ class InMemoryKey {
     var parts = key.split("://");
     this.protocol = parts[0];
     assert(this.protocol == 'in-memory');
-    parts = parts[1] ? parts[1].split('^^') : [];
+    parts = parts[1] ? parts.slice(1).join('://').split('^^') : [];
     this.arcId = parts[0];
     this.location = parts[1];
+    assert(this.toString() == key);
   }
   toString() {
-    return `${this.protocol}://${this.arcId}^^${this.location}`;
+    if (this.location !== undefined && this.arcId !== undefined)
```

Comment:
maybe worth adding a unit test for this?

Reply:
I'll think about testing for the key objects. A downside would be that we'd need to expose them - at the moment they're entirely hidden away inside the storage implementations.

Maybe we can do a separate storage-directory export though, and let the tests just import individual files directly.

Tracked in https://github.com/PolymerLabs/arcs/issues/587.